### PR TITLE
Optionally use `pluginName` property instead of constructor name

### LINKED ISSUE #6: Alternative to constructor name
We're creating wrapped plugins to support incremental builds in angular2, and quick-temp produces very poor names for these.

In ES6, Function.prototype.name is explicitly configurable per section 9.2.11 --- but in legacy engines (including old V8) this is not necessarily true. So, the temp directory name always looks like `diffing_plugin_wrapper` --- which is not super ideal, and will throw if we try to configure it unless we use a super bleeding edge version of V8/io.js.

As an alternative, it would be really great if there was an optional other way to do this.

I'm picturing something like this:

```js
class DiffingPluginWrapper {
  get pluginName() { return this.pluginClass.name; }
  // ...
}
```

```js
function prettyTmpDirName (obj, prop) {
  function cleanString (s) {
    return underscoreString.underscored(s || '')
      .replace(/[^a-z_]/g, '')
      .replace(/^_+/, '')
  }

  var cleanObjectName = (typeof obj.pluginName === 'string'
      ? obj.pluginName
      : (obj.constructor && obj.constructor.name)) || '';
  if (cleanObjectName) cleanObjectName += '-'
  var cleanPropertyName = cleanString(prop)
  return cleanObjectName + cleanPropertyName + '-XXXXXXXX.tmp'
}
```

### PR DESCRIPTION
**Pull request by caitp**
Closes #6

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -60,8 +60,12 @@ function prettyTmpDirName (obj, prop) {
       .replace(/^_+/, '')
   }
 
-  var cleanObjectName = cleanString(obj.constructor && obj.constructor.name)
-  if (cleanObjectName === 'object') cleanObjectName = ''
+  var pluginName = obj.pluginName;
```
Comment (caitp) [inline index.js]: cache here so that it's observably loaded only once

File: index.js [hunk 2]
```diff
@@ -60,8 +60,12 @@ function prettyTmpDirName (obj, prop) {
...
+  var cleanObjectName =
+      cleanString(typeof pluginName === 'string'
```
Comment (caitp) [inline index.js]: could `toString()` it but it doesn't seem valuable

File: index.js [hunk 3]
```diff
@@ -60,8 +60,12 @@ function prettyTmpDirName (obj, prop) {
...
+          ? pluginName
+          : (obj.constructor && obj.constructor.name));
```
Comment (caitp) [inline index.js]: I'm not sure where the dislike for semicolons comes from, but if it's necessary, will take them out

**Comment by caitp** [comment]
self-reviewed, and tested working

**Comment by joliss** [comment]
`.pluginName` would be a new piece of API, but we're already using `.description` in a bunch of places. How would you feel about using `.description` for the tmp dir name, if available?

Cc @stefanpenner, @rwjblue - Do you have objections to merging this (with `s/pluginName/description/`)? Also, if you have any thoughts, I'd be much interested to hear how you think Broccoli core might generate tmp dir names, once the new API lands and tmp dir handling is centralized in Broccoli core. It seems like there should be a neater solution than these types of heuristics.

File: index.js (hunk 3 above)
Comment (joliss) [inline index.js]: I just prefer editing semicolon-less code, so all my modules don't use semicolons, but I don't feel very strongly about it. ;-)

**Comment by joliss** [comment]
From #6, re the use case motivating this:

> We're creating wrapped plugins to support incremental builds in angular2, and quick-temp produces very poor names for these.

Do you have any of this wrapping code on GitHub? I'd like to understand the use case in more detail, so we can be sure that we support it properly.

**Comment by caitp** [comment]
The wrapper class we're using (which still has some bugs being hashed out) is at https://github.com/angular/angular/blob/master/tools/broccoli/diffing-broccoli-plugin.ts

**Comment by joliss** [comment]
Thanks, will take a look at the wrapper.

> "description" reads like a more verbose explanation of a tree rather than a short name

Yeah, I agree. I'm working on the new Broccoli API right now and need a feature like this as well. Let me mull it over for a bit.

**Comment by joliss** [comment]
I added this in 9501f25 as a third argument, e.g. `quickTemp.makeOrRemake(this, 'tmpDestDir', 'TreeMerger')`. Thanks again!

**Pull request closed without merging**